channel is not writable

```
2022-08-09 18:59:15.925 [ERROR] [tracing-sender] [org.bithon.agent.dispatcher.brpc.BrpcTraceMessageChannel] - [bTxId:, bSpanId:] Failed to send tracing: Failed to invoke ITraceCollector#sendTrace due to channel is not writable
```

Might be resolved by batching the tracing logs

--- a/component/component-brpc/src/main/java/org/bithon/component/brpc/channel/ClientChannel.java
+++ b/component/component-brpc/src/main/java/org/bithon/component/brpc/channel/ClientChannel.java
@@ -21,7 +21,7 @@ import org.bithon.component.brpc.endpoint.EndPoint;
 import org.bithon.component.brpc.endpoint.IEndPointProvider;
 import org.bithon.component.brpc.endpoint.SingleEndPointProvider;
 import org.bithon.component.brpc.exception.CallerSideException;
-import org.bithon.component.brpc.exception.ServiceInvocationException;
+import org.bithon.component.brpc.exception.ChannelException;
 import org.bithon.component.brpc.invocation.ServiceStubFactory;
 import org.bithon.component.brpc.message.in.ServiceMessageInDecoder;
 import org.bithon.component.brpc.message.out.ServiceMessageOutEncoder;
@@ -118,7 +118,7 @@ public class ClientChannel implements IChannelWriter, Closeable {
     public void writeAndFlush(Object obj) {
         Channel ch = channel.get();
         if (ch == null) {
-            throw new ServiceInvocationException("Client channel is closed");
+            throw new ChannelException("Client channel is closed");
         }
         ch.writeAndFlush(obj);
     }
@@ -137,7 +137,7 @@ public class ClientChannel implements IChannelWriter, Closeable {
     public synchronized void disconnect() {
         Channel ch = channel.getAndSet(null);
         if (ch != null) {
-            ch.disconnect();
+            ch.flush().disconnect();
         }
         connectionTimestamp = 0;
     }

--- a/component/component-brpc/src/main/java/org/bithon/component/brpc/channel/ServiceMessageChannelHandler.java
+++ b/component/component-brpc/src/main/java/org/bithon/component/brpc/channel/ServiceMessageChannelHandler.java
@@ -104,6 +104,19 @@ public class ServiceMessageChannelHandler extends ChannelInboundHandlerAdapter {
         }
     }
 
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        if (ctx.channel().isWritable()) {
+            // set auto read to true if channel is writable.
+            ctx.channel().config().setAutoRead(true);
+        } else {
+            log.warn("channel is not writable, disable auto reading for back pressing");
+            ctx.channel().config().setAutoRead(false);
+            ctx.flush();
+        }
+        ctx.fireChannelWritabilityChanged();
+    }
+
     public boolean isChannelDebugEnabled() {
         return channelDebugEnabled;
     }

--- a/component/component-brpc/src/main/java/org/bithon/component/brpc/exception/ChannelException.java
+++ b/component/component-brpc/src/main/java/org/bithon/component/brpc/exception/ChannelException.java
@@ -1,0 +1,27 @@
+/*
+ *    Copyright 2020 bithon.org
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package org.bithon.component.brpc.exception;
+
+/**
+ * @author frank.chen021@outlook.com
+ * @date 2022/8/10 21:33
+ */
+public class ChannelException extends ServiceInvocationException {
+    public ChannelException(CharSequence message) {
+        super(message);
+    }
+}

--- a/component/component-brpc/src/main/java/org/bithon/component/brpc/invocation/ClientInvocationManager.java
+++ b/component/component-brpc/src/main/java/org/bithon/component/brpc/invocation/ClientInvocationManager.java
@@ -20,6 +20,7 @@ import org.bithon.component.brpc.ServiceConfiguration;
 import org.bithon.component.brpc.channel.IChannelWriter;
 import org.bithon.component.brpc.exception.CalleeSideException;
 import org.bithon.component.brpc.exception.CallerSideException;
+import org.bithon.component.brpc.exception.ChannelException;
 import org.bithon.component.brpc.exception.TimeoutException;
 import org.bithon.component.brpc.message.in.ServiceResponseMessageIn;
 import org.bithon.component.brpc.message.out.ServiceRequestMessageOut;
@@ -112,7 +113,17 @@ public class ClientInvocationManager {
         if (debug) {
             //log.info("[DEBUGGING] Sending message: {}", message);
         }
-        channelWriter.writeAndFlush(serviceRequest);
+
+        for (int i = 0; i < 3; i++) {
+            try {
+                channelWriter.writeAndFlush(serviceRequest);
+                break;
+            } catch (ChannelException e) {
+                if (i < 2) {
+                    channelWriter.connect();
+                }
+            }
+        }
 
         if (inflightRequest != null) {
             try {

